src/proxy.ts

```diff
@@ -0,0 +1,109 @@
+import { fork, Sequence } from 'effection';
+import { AddressInfo } from 'net';
+import * as proxy from 'http-proxy';
+import * as http from 'http';
+import { listen, ReadyCallback } from './http';
+import { EventEmitter, forkOnEvent, resumeOnEvent } from './util';
+import * as trumpet from 'trumpet';
+import * as zlib from 'zlib';
+
+interface ProxyOptions {
+  port: number,
+  targetPort: number
+  inject?: string
+};
+
+function pipe(from, to) {
+  fork(function*() {
+    let writer = forkOnEvent(from, 'data', function*(chunk) {
+      to.write(chunk);
+    });
+
+    try {
+      yield resumeOnEvent(from, 'end');
+      to.end();
+    } finally {
+      writer.halt();
+    }
+  });
+}
+
+export function* createProxyServer(options: ProxyOptions, ready: ReadyCallback = x => x): Sequence {
+  let proxyServer = proxy.createProxyServer({
+    target: `http://localhost:${options.targetPort}`,
+    selfHandleResponse: true
+  });
+
+  forkOnEvent(proxyServer, 'proxyRes', function*(proxyRes, req, res) {
+    for(let [key, value] of Object.entries(proxyRes.headers)) {
+      res.setHeader(key, value);
+    }
+
+    let contentType = proxyRes.headers['content-type'] as string;
+    let contentEncoding = proxyRes.headers['content-encoding'] as string;
+
+    if(contentType && contentType.split(';')[0] === 'text/html') {
+      res.removeHeader('content-length');
+      res.removeHeader('content-encoding');
+
+      res.writeHead(proxyRes.statusCode, proxyRes.statusMessage);
+
+      let tr = trumpet();
+
+      tr.select('head', (node) => {
+        let rs = node.createReadStream();
+        let ws = node.createWriteStream();
+        ws.write(options.inject || '');
+        rs.pipe(ws);
+      });
```

Comment:
This callback will have to be converted to Effection, but how? We need to make sure that it doesn't leak when the request finishes, especially since this will be called quite often.

Reply:
I'm not sure, other than it should get cleaned up if th `tr.close()` method is called, but it seems like that's part of the issue.

Reply:
I came up with a solution to this which doesn't require any changes in Effection, check it out!

Reply:
Changed my mind on this after reading the stream documentation a bit.

Reply:
Still curious to hear what this was as it may be applicable later 😎

Reply:
You can see it here: https://github.com/bigtestjs/server/blob/72fb9b758502f1dd70ec7a5e3beb41b36fe5e7c0/src/proxy.ts